### PR TITLE
RA-1711: Program Enrollment Widget should allow enrollment date of program to be equal to completion date of previous program

### PR DESCRIPTION
**Pull request by mogoodrich**

### CODE REVIEW

File: omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/program/ProgramHistoryFragmentController.java [hunk 1]
```diff
@@ -75,8 +76,48 @@ public void controller(FragmentConfiguration config, @FragmentParam("app") AppDe
...
+        Collections.sort(patientPrograms, new Comparator<PatientProgram>() {
+            @Override
+            public int compare(PatientProgram patientProgram1, PatientProgram patientProgram2) {
+
+                if (patientProgram1.getDateEnrolled().after(patientProgram2.getDateEnrolled())) {
+                    return -1;
+                }
+                else if (patientProgram1.getDateEnrolled().before(patientProgram2.getDateEnrolled())) {
+                    return 1;
+                }
```
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/program/ProgramHistoryFragmentController.java:88]: For the above, could use OpenmrsUtil.compareWithNullAsEarliest(patientProgram1.getDateEnrolled(), patientProgram2.getDateEnrolled()), and return it if non-zero.

File: omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/program/ProgramHistoryFragmentController.java [hunk 2]
```diff
@@ -75,8 +76,48 @@ public void controller(FragmentConfiguration config, @FragmentParam("app") AppDe
...
+                // having two programs with the same start date but different end dates should be illegal, but just in case
+                else if (patientProgram1.getDateCompleted().after(patientProgram2.getDateCompleted())) {
+                    return -1;
+                }
+                else if (patientProgram1.getDateCompleted().before(patientProgram2.getDateCompleted())) {
+                    return 1;
+                }
+
```
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/program/ProgramHistoryFragmentController.java:105]: For the above, could use OpenmrsUtil.compareWithNullAsLatest(patientProgram1.getDateCompleted(), patientProgram2.getDateCompleted()), and return it if non-zero.

File: omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/program/ProgramHistoryFragmentController.java [hunk 3]
```diff
@@ -75,8 +76,48 @@ public void controller(FragmentConfiguration config, @FragmentParam("app") AppDe
...
+                else if (patientProgram1.getDateCreated().after(patientProgram2.getDateCreated())) {
+                    return -1;
+                }
+                else if (patientProgram1.getDateCreated().before(patientProgram2.getDateCreated())) {
+                    return 1;
+                }
```
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/program/ProgramHistoryFragmentController.java:111]: For the above, could use OpenmrsUtil.compare(patientProgram1.getDateCreated(), patientProgram2.getDateCreated()), and return it if non-zero.

File: omod/src/main/web/dashboardwidgets/programstatus/programstatus.controller.js [hunk 4]
```diff
@@ -744,4 +747,58 @@ export default class ProgramStatusController {
...
+
+        if (dateCreated1 > dateCreated2) {
+          return -1;
+        }
+        else if (dateCreated1 < dateCreated2) {
+          return 1;
+        }
+
+        // and finally, just use uuid, because we need to make sure the sort order is deterministic
+        if (patientProgram1.value.uuid < patientProgram2.value.uuid) {
```
Comment (mseaton) [inline omod/src/main/web/dashboardwidgets/programstatus/programstatus.controller.js:796]: Shouldn't this be consistent with what is in the java code?  Either both use "id" or both use "uuid"?

**Review by mseaton — APPROVED** [review]
Fine with me - a few comments.

**Comment by dkayiwa** [comment]
If they do not exist already, adding some tests for the comparator would be great. 😊

**Comment by mogoodrich** [comment]
@dkayiwa  fair enough, re: tests

**Comment by mogoodrich** [comment]
Added tests of the comparator in Java... ran into a bit of a rabbit hole doing tests in JavaScript as it turns out that PhantomJS doesn't _quite_ handle dates the same way Chrome does... going to merge this in as-is for now...

**Pull request merged**